core/ir/x86/encode.c

```diff
@@ -1284,17 +1284,7 @@ opnd_type_ok(decode_info_t *di /*prefixes field is IN/OUT; x86_mode is IN*/, opn
         return (opnd_is_far_pc(opnd) || opnd_is_far_instr(opnd));
     }
     case TYPE_O:
-        return ((opnd_is_abs_addr(opnd) ||
-#ifdef X64
-                 /* We'll take a relative address that rip-rel won't reach:
-                  * after all, OPND_CREATE_ABSMEM() makes a rip-rel.
-                  */
-                 (opnd_is_rel_addr(opnd) &&
-                  (!REL32_REACHABLE(di->final_pc + MAX_INSTR_LENGTH,
```

Comment:
> di->final_pc is NULL (not initialized) when called from encoding_possible(). So the reachability check does not really work and it's not necessary to determine possible encodings.

But di->final_pc is set by instr_encode_arch() right before calling encoding_possible().  That is the typical call path.  The instance where final_pc is not set is when called by get_encoding_info() for e.g. instr_zeroes_ymmh().

Reply:
So this is used in walking the encoding template possibilities and I'm guessing TYPE_O is early in the sequence, as it is more specialized.  That's why it needs to exclude the reachable so that will bypass and pick a later template.

So the two calls outside a regular encode, get_encoding_info() and instr_is_encoding_possible(), which don't take in a final PC, may pick the wrong template for MOV.  Although it won't matter for instr_is_encoding_possible() who just wants a yes/no answer.  For the various callers of get_encoding_info(): instr_shrink_to_16_bits() is the only one which might have an issue: but it looks like it's ok.

I think what we want to do is document this and add comments in each caller stating that they're fine with this caveat.

Reply:
I created #6202 for this.  Could I ask you to test it in your ubsan setup to confirm it eliminates the problem in front of you?

Reply:
Thanks! Let's move the conversation to that PR.